### ISSUE TITLE
Cannot flush DB when user preference model app is not installed

### ISSUE THREAD

**Issue by leetucker** [EliotBerriot/django-dynamic-preferences#83]
If I try to run `python manage.py flush` with the global dynamic preferences app installed, but the per-user preferences app not installed, I get the following error:

```
CommandError: Database blhdjango couldn't be flushed. Possible reasons:
  * The database isn't running or isn't configured correctly.
  * At least one of the expected database tables doesn't exist.
  * The SQL was invalid.
Hint: Look at the output of 'django-admin sqlflush'. That's the SQL this command wasn't able to run.
The full error: cannot truncate a table referenced in a foreign key constraint
DETAIL:  Table "dynamic_preferences_users_userpreferencemodel" references "auth_user".
HINT:  Truncate table "dynamic_preferences_users_userpreferencemodel" at the same time, or use TRUNCATE ... CASCADE.
```

My guess is that the issue arises because the user preference model tables are created and then moved in the global app's migrations, even though the table is never used in the global preferences app.  Is there a way to fix this so that I don't need to go through the process of constructing manual SQL code to clear the database every time that I want to load new data into my dev instance?

**Comment by EliotBerriot** [comment]
Thanks for reporting that!

If I understand the issue properly:

1. You have a fresh projects
2. You apply migrations
3. Because of how we handled things before in dynamic_preferences, the migrations create a UserPrefrence model then moves it to another table
4. You try to flush the database, but since the dynamic_preferences.users app is not installed, it fails

This is indeed serious, though I'm not sure how to fix this while keeping backward compatibility.

My first idea was to apply the user migrations (in 0001 and 0004) only if `dynamic_preferences.users` is present in installed apps but this would not work as the `users.0001` migrations relies on existing model being created in those migrations.

I'm a bit stuck now and I'm open to suggestions for solving this problem :)

**Comment by leetucker** [comment]
I'm not an expert on migrations, so this may not work, but it seems worth suggesting.

This is my understanding of the current situation.  If a user currently has installed the core app with version <1.0, then they have the UserPreference model table.  If they have the core app installed with version >=1.0, then the table is renamed, the underlying model is deleted, and if and when the user migration 0001 is applied, the preexisting data is found and used

So, though this might not address all cases, it seems to be that you could perhaps apply the following which will address the most common use case:

- In core migration 0004, rename the users table if (and only if) it is non-empty, otherwise, simply delete the table.  This does not lead to any data loss, but it does mean that there is no table inconsistency in any situation where the user preference model was never actually used
- In users migration 0001, create the users table if it doesn't already exist, while still loading the existing UserPreferenceModel data if it exists.  This table can still retain the dependency on core migration 0004, and user preference data will be retained whenever it exists.

I'm not sure if the conditional logic above would create an issue for someone who wanted to regress back from a version >1.0 to a version <1.0, but perhaps that isn't a large concern given that you're now several subversions ahead of this.

**Comment by EliotBerriot** [comment]
I like your suggestion and I see no obvious pitfall with this approach. I don't know how we can apply migrations operations conditionally (based on a table presence or absence), but I'd be suprised if they were no way to do it.

Thanks for your time :)

**Comment by EliotBerriot** [comment]
Fixed in #142
